packages/vite/src/node/plugins/css.ts

```diff
@@ -1949,101 +2024,168 @@ function fixScssBugImportValue(
+      ) => {
+        // eslint-disable-next-line no-restricted-globals
+        const sass: typeof Sass = require(sassPath)
+        // eslint-disable-next-line no-restricted-globals
+        const path = require('node:path')
+
+        // NOTE: `sass` always runs it's own importer first, and only falls back to
+        // the `importer` option when it can't resolve a path
+        const _internalImporter: Sass.Importer = (url, importer, done) => {
+          internalImporter(url, importer, options.filename).then(
+            (data) => done?.(data),
+          )
         }
-      : {}),
-  }
-
-  try {
-    const result = await new Promise<Sass.Result>((resolve, reject) => {
-      render(finalOptions, (err, res) => {
-        if (err) {
-          reject(err)
-        } else {
-          resolve(res)
+        const importer = [_internalImporter]
+        if (options.importer) {
+          Array.isArray(options.importer)
+            ? importer.unshift(...options.importer)
+            : importer.unshift(options.importer)
         }
-      })
-    })
-    const deps = result.stats.includedFiles.map((f) => cleanScssBugUrl(f))
-    const map: ExistingRawSourceMap | undefined = result.map
-      ? JSON.parse(result.map.toString())
-      : undefined
 
-    return {
-      code: result.css.toString(),
-      map,
-      additionalMap,
-      deps,
-    }
-  } catch (e) {
-    // normalize SASS error
-    e.message = `[sass] ${e.message}`
-    e.id = e.file
-    e.frame = e.formatted
-    return { code: '', error: e, deps: [] }
-  }
-}
-
-const sass: SassStylePreprocessor = (source, root, options, aliasResolver) =>
-  scss(
-    source,
-    root,
+        const finalOptions: Sass.Options = {
+          ...options,
+          data,
+          file: options.filename,
+          outFile: options.filename,
+          importer,
+          ...(options.enableSourcemap
+            ? {
+                sourceMap: true,
+                omitSourceMapUrl: true,
+                sourceMapRoot: path.dirname(options.filename),
+              }
+            : {}),
+        }
+        return new Promise<{
+          css: string
+          map?: string | undefined
+          stats: Sass.Result['stats']
+        }>((resolve, reject) => {
+          sass.render(finalOptions, (err, res) => {
+            if (err) {
+              reject(err)
+            } else {
+              resolve({
+                css: res.css.toString(),
+                map: res.map?.toString(),
+                stats: res.stats,
+              })
+            }
+          })
+        })
+      },
     {
-      ...options,
-      indentedSyntax: true,
+      parentFunctions: { internalImporter },
+      shouldUseFake(_sassPath, _data, options) {
+        return !!(
+          (options.functions && Object.keys(options.functions).length > 0) ||
+          (options.importer &&
+            (!Array.isArray(options.importer) || options.importer.length > 0))
+        )
+      },
```

Comment:
IIUC we use a fake worker / run in main thread when non-serializable options are passed? Could we add a comment here, and perhaps a debug message that we're falling back to running in main thread? (Same for other workers)

Reply:
Yes, that's right. I added a comment 👍